docs/GettingStartedDocs/best_practices_secure_enclaves.md

```diff
@@ -0,0 +1,275 @@
+
+```diff
+int ecall_with_user_check4(void* ptr) {
+
+    blob_t* blob;
+    void* volatile captured_data;
+    volatile size_t captured_size;
+
+    // Ensure passed-in pointer is not null and the memory block is located entirely outside of the enclave.
+    if (!oe_is_outside_enclave(ptr, sizeof(blob_t)))
+        return -1;
+
+    blob = (blob_t*)ptr;
+
+    // Capture blob descriptors to avoid TOCTOU problems.
+    captured_data = blob->data;
+    captured_size = blob->size;
+
+    // Blobs to be rendered are limited to 4K.
+    if (captured_size > 4096)
+        return -1;
+
++   // Ensure validity of nested structure
++   if (!oe_is_outside_enclave(captured_data, captured_size))
++       return -1;
+
++   // Data parsed by render_data is "flat", non-self-referencing.
+    return render_data(captured_data, captured_size);
+}
+```
+## 1.5 Corrected enclave ecall with custom marshaling
+
+Thanks for sticking with us. Here is our corrected ecall function that marshals the parameter correctly:
+```c++
+int ecall_with_user_check5(void* ptr) {
+
+    blob_t* blob;
+    void* volatile captured_data;
+    volatile size_t captured_size;
+
+    // Ensure passed-in pointer is not null and the memory block is located entirely outside of the enclave.
+    if (!oe_is_outside_enclave(ptr, sizeof(blob_t)))
+        return -1;
+
+    blob = (blob_t*)ptr;
+
+    // Capture blob descriptors to avoid TOCTOU problems.
+    captured_data = blob->data;
+    captured_size = blob->size;
+
+    // Blobs to be rendered are limited to 4K.
+    if (captured_size > 4096)
+        return -1;
+
+    // Ensure validity of nested structure
+    if (!oe_is_outside_enclave(captured_data, captured_size))
+        return -1;
+
+    // Data parsed by render_data is "flat", non-self-referencing.
+    return render_data(captured_data, captured_size);
+}
+```
+
+<br />
+<br />
+<br />
+
+# 2. Handling Secrets in Enclave Applications
+
+Open Enclave is an SDK that helps developers build apps that will run inside a hardware-based Trusted Execution Environment (TEE). At their core, TEEs protect application code and data at runtime from the host environment. Without new, hardware-implemented protections, a malicious or compromised host operating system would be able to modify code or read data. Any secrets managed by an application during runtime would be at risk of exposure. Open Enclave SDK helps developers build applications that are protected from these threats.
+
+> It's important to note that enclaves do not provide _code confidentiality_. This is noteworthy in the context of this discussion because, as will be discussed, hard-coded credentials in application code are to be avoided -- with or without enclave protections.
```

Comment:
```suggestion
> It's important to note that enclaves do not generally provide _code confidentiality_. This is noteworthy in the context of this discussion because, as will be discussed, hard-coded credentials in application code are to be avoided -- with or without enclave protections.
```

Add "generally" since there are ways to do it by loading an encrypted code module into an enclave.

Reply:
Added "generally".